Patch to avoid accidental assert failures from EnvT::KeywordIx

maynardGK opened the pull request:
Converts a hard exit/crash to a harmless diagnostic for new imperfect software. (see comments i n commit). Also removal of a modification from 2014 that was suppose to prevent duplicate keywords.

Greg Jung 
gvjung@gmail.com

maynardGK committed:
Patch to avoid accidental assert failures upon keyword lookup.

  To streamline keyword processing we use a trick to look up once and for all the keyword index.
 For almost all keywords, we use
    static keywordnameIx = e->KeywordIx("KEYWORDNAME");
 where the call from the interpreter is given by
	 BaseGDL* function_fun(EnvT* e)  { ...
 or by
                void routine_pro(EnvT *e) { ...
   The indexed list is created in the corresponding call to DLibFun, DLibPro,
DLibFunRetNew etc. as found in the libinit*.cpp files.  The routine EnvT::KeywordIx finds the
index from the provided string.  If there is no corresponding string, pro->FindKey()
returns instead a value of -1.  Through version 0.9.8, the result "-1" is tested with an
assert( val != -1) leading to program exit in an undignified manner.

  This commit is to change that behavior so that, instead of assert() being invoked,
a diagnostic message is issued (3 lines, including info on procedure call and keyword intended).
Then, instead of throwing an exception, the index value is set to zero.  So that accidental
usage of the uninitialized keyword will be interpreted as the first keyword.

Greg Jung (aka maynardGK)
gvjung@gmail.com

maynardGK committed:
Minor corrective patches to envt.cpp

maynardGK commented:
The error that stopped the test was an internal server error. The rest look ok.

slayoo commented:
I've just restarted the failed job

GillesDuvert commented:
Hi @maynardGK , I do understand the intent, I had once wanted to get rid of these annoying asserts.
But consider: this particular assert serves to find a severe programming bug that should be cured asap. **As asserts are not trapped in Releases**, the presence of such programming errors is invisible for normal users (all the GetKw() functions return NULL as if the Kw was absent). It is thus mandatory to keep a way to inform the authors that they forgot to enter the keyword in the dictionary. We as authors must test our changes in Debug mode first and with these asserts we are bound to find our mistakes, sooner or later.
Some heavily used programs trap such errors and ask users to report by sending an automatic bug report. We are not prepared to do that I believe.

maynardGK commented:
@GillesDuvert 
That's ok, but prior to exit the message should be made - or the bug remains undiagnosed.
Setting the index to zero - thereby invoking the first possible keyword - may be too lax, I agree.
But the message is helpful, and a message with exception, preventing execution of the subroutine,
may be enough of a "heads up".  I'll commit on top of this with an exception, then.

maynardGK committed:
Issue diagnotic message, proceed with assert.

  The assert is put back in place for this condition but a diagnositc messsage
is given prior to the assert().

maynardGK commented:
I am closing this as the latest modification is used in basic_fun branch to restart testing.
